web3/types.py

```diff
@@ -0,0 +1,122 @@
+from typing import (
+    Any,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    Union,
+)
+
+from eth_typing import (
+    BlockNumber,
+    ChecksumAddress,
+    Hash32,
+    HexStr,
+)
+from hexbytes import (
+    HexBytes,
+)
+
+from web3._utils.compat import (
+    TypedDict,
+)
+
+# todo: move these to eth_typing once web3 is type hinted
+
+
+ABIEventParameter = TypedDict("ABIEventParameter", {
+    "name": str,
+    "type": str,
+    "indexed": bool,
+})
+
+
+ABIEvent = TypedDict("ABIEvent", {
+    "type": "event",
+    "name": str,
+    "inputs": Sequence[ABIEventParameter],
+    "anonymous": bool,
+})
+
+
+ABIFunctionComponents = TypedDict("ABIFunctionParameter", {
+    "name": str,
+    "type": str,
+    "components": Optional[Sequence["ABIFunctionComponents"]],
+})
+
+
+ABIFunctionParameter = TypedDict("ABIFunctionParameter", {
+    "name": str,
+    "type": str,
+    "components": Optional[Sequence[ABIFunctionComponents]],
+})
+
+
+# should we get more granular and define constructor / fallback individually?
+ABIFunction = TypedDict("ABIFunction", {
+    "type": Union["function", "constructor", "fallback"],
+    "name": str,
+    "inputs": Sequence[ABIFunctionParameter],
+    "outputs": Optional[Sequence[ABIFunctionParameter]],
+    "stateMutability": Union["pure", "view", "nonpayable", "payable"],
+    "payable": bool,
+    "constant": bool,
+})
+
+
+ABI = Sequence[Union[ABIFunction, ABIEvent]]
+
+BlockIdentifier = Union[str, BlockNumber, bytes, Hash32]
```

Comment:
I think `bytes` is out of place since `Hash32` is intended to represent a 32-byte `bytes` value.